Core/clim-basic/repaint.lisp

```diff
@@ -26,23 +26,55 @@
 
 (in-package :clim-internals)
 
+;;; Delayed repainting mechanism
+;;;
+;;; A region of code the execution of which will result in multiple
+;;; redundant repaints of one or more sheets can be surrounded in
+;;; [CALL-]WITH-DELAYED-DISPATCH-REPAINT. In that case, repaints
+;;; dispatched in the body code are not processed immediately but are
+;;; recorded for later. When exiting from the
+;;; [CALL-]WITH-DELAYED-DISPATCH-REPAINT call, all recorded repaints
+;;; are merged (i.e. for each involved sheet, the effective repaint
+;;; region is the union of all recorded repaint regions) and executed.
+
+(defvar *delayed-repaints* nil)
+
+(defmethod dispatch-repaint :around ((sheet basic-sheet) region)
+  ;; If *DELAYED-REPAINTS* is non-NIL, push REGION onto the repaint
+  ;; queue for SHEET instead of dispatching the repaint. Otherwise,
+  ;; just call the next method.
+  (if-let ((delayed-repaints *delayed-repaints*))
+    (push region (gethash sheet delayed-repaints '()))
+    (call-next-method)))
+
+(defun call-with-inhibited-dispatch-repaint (thunk)
+  ;; If *DELAYED-REPAINTS* is non-NIL, there must be a surrounding
+  ;; call of this function, so we don't have to do anything.
+  (if *delayed-repaints*
+      (funcall thunk)
+      (progn
+        (let ((delayed-repaints (make-hash-table :test #'eq)))
+          ;; DISPATCH-REPAINT calls in THUNK populate
+          ;; DELAYED-REPAINTS.
+          (let ((*delayed-repaints* delayed-repaints))
+            (funcall thunk))
+          ;; Merge and execute recorded repaint requests.
+          (maphash (lambda (sheet regions)
+                     (dispatch-repaint sheet (reduce #'region-union regions)))
+                   delayed-repaints)))))
+
+(defmacro with-inhibited-dispatch-repaint (() &body body)
+  `(call-with-inhibited-dispatch-repaint (lambda () ,@body)))
+
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;;
 ;;; Repaint protocol functions.
 
-(defmethod dispatch-repaint ((sheet graft) region)
+#+no (defmethod dispatch-repaint :around ((sheet graft) region)
   (declare (ignore sheet region)))
```

Comment:
I suppose this `#+no` means, that the method definition is meant to be removed?

Reply:
Removed.